vllm/core/scheduler.py

```diff
@@ -1584,64 +1636,174 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
-                # not actually support chunking prompts.
-                #
-                # Therefore, `num_new_tokens` is computed in the same fashion
-                # for both multi-step+chunked-prefill &
-                # multi-step+chunked-prefill+APC
-                #
-                # Prompts with more tokens than the current remaining budget
-                # are postponed to future scheduler steps
-                if num_new_tokens > self._get_prompt_limit(seq_group):
-                    # If the seq_group is in prompt-stage, pass the
-                    # num_new_tokens as-is so the caller can ignore
-                    # the sequence.
-                    pass
-                else:
-                    num_new_tokens = 0 \
-                        if num_new_tokens > remaining_token_budget \
-                        else num_new_tokens
-            elif self.cache_config.enable_prefix_caching:
-                # When prefix caching is enabled, we always allocate
-                # the number of new tokens that is dividable by the block
-                # size to avoid partial block matching.
-                block_size = self.cache_config.block_size
-                remainder = budget.token_budget % block_size
-                if remainder != 0:
-                    raise ValueError("When enabling chunked prefill and "
-                                     "prefix caching, max_num_batched_tokens "
-                                     "(chunk size) must be dividable by "
-                                     "block size, but got chunk_size "
-                                     f"({budget.token_budget}) % block_size "
-                                     f"({block_size}) = {remainder}")
-                if remaining_token_budget < num_new_tokens:
-                    num_new_tokens = (remaining_token_budget //
-                                      block_size) * block_size
+            # Chunk if a running request cannot fit in the given budget.
+            # If number of seq > 1, it means it is doing beam search
+            # in a decode phase. Do not chunk.
+            num_uncached_new_tokens = self._chunk_new_tokens_to_schedule(
+                budget,
+                self._get_prompt_limit(seq_group),
+                num_uncached_new_tokens,
+            )
+
+        return num_uncached_new_tokens, num_cached_new_tokens
+
+    def _get_num_new_tokens_uncached_and_cached(
+            self, seq: Sequence) -> Tuple[int, int]:
+        """
+        Returns the number of new uncached and cached tokens for a given
+        sequence.
+
+        Returns:
+            A tuple of two ints. The first int is the number of uncached tokens.
+            The second int is the number of cached tokens.
+        """
+        if not seq.is_prefill():
+            # Decode sequences should always just have 1 uncached token
+            # TODO(rickyx): Actually is this still correct for multi-step?
+            return 1, 0
```

Comment:
And speculative decoding?

Reply:
hmm I might be missing something - but I think both speculative decoding and multi-step are using some `num_steps` field that's opaque to the scheduler as of now. 

So in a way, doesn't seem the shceduler is multi-step aware.